crates/bevy_ecs/src/storage/table.rs

```diff
@@ -83,7 +83,11 @@ impl TableRow {
     /// Creates a `TableRow`.
     #[inline]
     pub const fn new(index: usize) -> Self {
-        Self(index as u32)
+        let index_u32 = index as u32;
+        if index_u32 as usize != index {
```

Comment:
All of this could just be condensed into a single assert for `index`. You want to ensure the index can fit within a `u32`, then just check if `index` is greater than `u32::MAX`, and casting `u32` to `usize` will be fine since we also don't support 16-bit platforms or smaller.